src/Bicep.Core/Emit/TemplateWriter.cs

```diff
@@ -244,6 +249,197 @@ private void EmitParameter(JsonTextWriter jsonWriter, ParameterSymbol parameterS
             jsonWriter.WriteEndObject();
         }
 
+        private void EmitTypeDeclaration(JsonTextWriter jsonWriter, DeclaredTypeSymbol declaredTypeSymbol, ExpressionEmitter emitter)
+        {
+            jsonWriter.WriteStartObject();
+
+            var typeDefinitionObject = TypePropertiesForTypeExpression(declaredTypeSymbol.Type, declaredTypeSymbol.DeclaringType, declaredTypeSymbol.DeclaringType.Value);
+
+            foreach (var property in typeDefinitionObject.Properties)
+            {
+                if (property.TryGetKeyText() is string propertyName)
+                {
+                    emitter.EmitProperty(propertyName, property.Value);
+                }
+            }
+
+            jsonWriter.WriteEndObject();
+        }
+
+        private ObjectSyntax TypePropertiesForTypeExpression(TypeSymbol type, DecorableSyntax parentSyntax, SyntaxBase typeExpressionSyntax)
+            => AddDecoratorsToBody(parentSyntax, TypePropertiesForUndecoratedTypeExpression(type, parentSyntax, typeExpressionSyntax), type);
+
+        private ObjectSyntax TypePropertiesForUndecoratedTypeExpression(TypeSymbol type, DecorableSyntax parentSyntax, SyntaxBase typeExpressionSyntax)
```

Comment:
I don't really have a stronger objection to it than "it's kinda weird", but using the Bicep syntax tree as a means of writing JSON does feel strange.

Reply:
The decorator contract forces this approach at present, since any decorator can modify the compiled syntax and [expects that syntax to arrive as a Bicep object](https://github.com/Azure/bicep/blob/816a6dea7d24f4b035b3b38bd01ff83afc8622d6/src/Bicep.Core/Semantics/Decorator.cs#L17). We can change this, but it might be better to address as tech debt in another PR since it would require updating how every statement that can be decorated gets compiled.

Reply:
Thanks, that makes sense.